### DISCUSSION TITLE
Adjusting the Difficulty Adjust and Rate Adjust mods to more accurately emulating the functionality of OsuTrainer/McOsu

### DISCUSSION THREAD

**Discussion by Chadcel** [ppy/osu#27040] [Ideas]
[OsuTrainer](https://github.com/FunOrange/osu-trainer): An external utility to alter a beatmap's properties
[McOsu](https://store.steampowered.com/app/607260/McOsu/): An alternate client with many features to adjust a beatmap's properties

When I first saw the DA mod and the rate adjust changes to DT/NC/HT/DC, the first thought that came to my head was functionality similar to the above mentioned utilities. While they do accomplish this, they do so non-optimally.

DA imitates the behaviour of HR/EZ in that it's changes are applied before rate change mods. Suppose someone wanted to play a map with very heavy rate down with high AR. Values like 0.5x with AR10 are unachievable with this implementation. Furthermore, this causes a conflict between me thinking "Okay, I set it to AR9 because I want it to be AR9" and the actual AR being affected by the rate adjust mod I'm also playing with. I understand this can be bypassed by simply looking at the beatmap properties panel in the mod selection screen, but this seems like an unnecessary inconvenience.

Tangentially, rate change mods also work in a way that limits "trainer-like" behaviour. Currently, they are solely multipliers, emulating the behaviour of stable. Suppose I have a collection of beatmaps that are of varying BPMs, and I want them all to be exactly 180BPM. With the current implementation of rate adjust mods, I'd need to go and adjust my mod settings for every map, potentially even switching from DT/NC to HT/DC. This seems like further unnecessary inconvenience. Functionality to directly set a given beatmap's BPM rather than apply a multiplier to it would correct this.

Lastly, it'd be nice to see highly extended limits on DA and rate adjust mods to allow for esoteric values like ar13 or maps at 0.1x rate, but that might be a bit much even for a pedant like me.

**Comment by Rekunan** [comment]
DA-rate adjust interaction suggestion is duplicate of: #15249
BPM input suggestion is duplicate of: #22616

I feel like allowing esoteric values would be highly unnecessary: mcosu caps at ar12.5 and 0.5x and 2.0x is already too slow/fast for most maps.
